### PR TITLE
[SPARK-11121][Core] Correct the TaskLocation type

### PR DESCRIPTION
**Pull request by zhichao-li**
Correct the logic to return `HDFSCacheTaskLocation` instance when the input `str` is a in memory location.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/scheduler/TaskLocation.scala [hunk 1]
```diff
@@ -62,12 +62,5 @@ private[spark] object TaskLocation {
...
    */
-  def apply(str: String): TaskLocation = {
-    val hstr = str.stripPrefix(inMemoryLocationTag)
-    if (hstr.equals(str)) {
-      new HostTaskLocation(str)
-    } else {
-      new HostTaskLocation(hstr)
-    }
-  }
+  def apply(str: String): TaskLocation = new HostTaskLocation(str.stripPrefix(inMemoryLocationTag))
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/scheduler/TaskLocation.scala]: Seems OK, but this seems too minor to bother with the overhead of reviewing and testing. Maybe you can consider changes like this as part of larger logical changes in the future?

File: core/src/main/scala/org/apache/spark/scheduler/TaskLocation.scala [hunk 2]
```diff
@@ -62,12 +62,5 @@ private[spark] object TaskLocation {
    * These strings have the form [hostname] or hdfs_cache_[hostname], depending on whether the
    * location is cached.
    */
-  def apply(str: String): TaskLocation = {
-    val hstr = str.stripPrefix(inMemoryLocationTag)
-    if (hstr.equals(str)) {
-      new HostTaskLocation(str)
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/scheduler/TaskLocation.scala]: Should here be `HDFSCacheTaskLocation` in original mean?

Comment (zhichao-li) [inline core/src/main/scala/org/apache/spark/scheduler/TaskLocation.scala]: Good point, but I think it should be 
scala
```
if (hstr.equals(str)) // meaning not a inMemoryLocation
{
  new HostTaskLocation(str)
} else {  // inMemoryLocation
  new HDFSCacheTaskLocation(hstr)
}

```

**Comment by jerryshao** [comment]
Yeah, I misunderstood the return value of `stripPrefix`, LGTM.

**Comment by zhichao-li** [comment]
retest this please

**Comment by srowen** [comment]
@zhichao-li this needs a JIRA

**Comment by zhichao-li** [comment]
retest this please

**Comment by zhichao-li** [comment]
retest this please.

**Comment by jerryshao** [comment]
Do we need to verify this HDFS tag actually works with this modification, in case we don't mess the original implementation.

**Comment by zhichao-li** [comment]
Sounds good to have one. let me update the unit test.

File: core/src/main/scala/org/apache/spark/scheduler/TaskLocation.scala [hunk 3]
```diff
@@ -64,10 +64,10 @@ private[spark] object TaskLocation {
    */
   def apply(str: String): TaskLocation = {
     val hstr = str.stripPrefix(inMemoryLocationTag)
-    if (hstr.equals(str)) {
+    if (hstr.equals(str)) { // not a in memory location
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/TaskLocation.scala]: this comment doesn't make sense. I think you can just remove it.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala [hunk 4]
```diff
@@ -759,9 +759,9 @@ class TaskSetManagerSuite extends SparkFunSuite with LocalSparkContext with Logg
     val sched = new FakeTaskScheduler(sc,
         ("execA", "host1"), ("execB", "host2"), ("execC", "host3"))
     val taskSet = FakeTask.createTaskSet(3,
-      Seq(HostTaskLocation("host1")),
-      Seq(HostTaskLocation("host2")),
-      Seq(HDFSCacheTaskLocation("host3")))
+      Seq(TaskLocation("host1")),
+      Seq(TaskLocation("host2")),
+      Seq(TaskLocation("hdfs_cache_host3")))
```
Comment (andrewor14) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala]: can you add a finer-grained test case for this? e.g.
```
assert(TaskLocation("host1") === HostTaskLocation("host1"))
assert(TaskLocation("hdfs_cache_host3") === HDFSCacheTaskLocation("host3"))
```

**Comment by andrewor14** [comment]
@zhichao-li This looks fine, but are there other implications? In `TaskSetManager` there are entire blocks of code that are not executed before this patch.

**Comment by zhichao-li** [comment]
I think there would be no other hidden pitfall. Just grep the code, most of the HDFSCacheTaskLocation is created explicitly except  for (https://github.com/apache/spark/blob/f85aa06464a10f5d1563302fd76465dded475a12/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L1543) , which would not run into the case of in memory host.

**Comment by zhichao-li** [comment]
retest this please

**Comment by zhichao-li** [comment]
cc @srowen @andrewor14 seems it pretty close to be merged. any more comments?

**Comment by srowen** [comment]
It looks like this is a duplicate/subset of https://github.com/apache/spark/pull/9181 -- is that covering the same ground?

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**